### DISCUSSION TITLE
[feature request] Improve Include-Page tag to pull only specific content

### DISCUSSION THREAD

**Discussion by PeterMarkuson** [gollum/gollum#1832] [General]
I love Gollum for the include-tag feature (https://github.com/gollum/gollum/wiki#include-tag), because it allows me to assemble and re-use information I have already written down somehwere else. When updating the source of the included information, automatically all other pages using that information are up-to-date immediately 👍 

To be flexible and make full use of that feature, a lot of pages with only small content need to be created and then included in larger pages (in my case setup guides). So, I end up with many atomic pages and little content on each page. Many information blocks build-up on previous ones and would be good to have them on one page together (for editing, review). But, it makes not always sense to include all of it in another page. An improved include-tag with a parameter to select only a block of the page would be a great improvement.

My proposal:

* Today: User needs to create a page for every information he wants to re-use, because the include-tag feature always pulls the full page.
* Improvement: A user could specify a bookmark and in that case only that bookmark and the content below is included in the page.
* Result: Only the content from the bookmark is included. 

The content is identified either from bookmark to the next bookmark or from the level of the heading to the next same level.

**Comment by benjaminwil** [comment]
Thanks for submitting this request.

Based on how `[[include]]` is currently implemented, I think what you have proposed would greatly increase the complexity of the existing feature. If we were to implement this, I think it would have to be a new, separate feature.

Just going to write out my thoughts so far below.

I'm also not sure practically how we'd implement this from a user experience point of view. I think that the proposed feature could be more error prone for writers and editors. Right now the `[[include]]` cares very little about the content of the included page. In the proposed feature we now must parse many additional "nodes" and match the right one(s).

I personally see the atomic stub articles that `[[include]]` forces a writer to manage as a benefit. While they can "pollute" the wiki file directory, I think that managing them is very straightforward overall, and a team of writers and editors will have no trouble backtracking from an `[[include]]` to a specific file with a specific path. I would be curious to hear more about why, in your case, the atomic stub article isn't ideal for you? I'm also curious whether a custom macro could get you closer to what your ideal workflow would be.

**Comment by dometto** [comment]
This is to elaborate a bit on why this would potentially be quite complex, even in a custom macro.

The feature would require the at least partial rendering of the included page. Without a rendered document, we do not have any headings (html `<h1>-<h6>` tags) to include. Rendering the included document *before* it is included is inefficient, but we could theoretically do it. However, when we have the rendered document, there's no easy way to get back to the corresponding line numbers in the source document. That would require the rendered HTML to include a "source map", which not all rendering gems for all markups support. This was the core difficulty behind implementing the [editable sections](https://github.com/gollum/gollum/wiki/5.0-release-notes#editable-sections), and we only solved it on the frontend with JavaScript.

Bottom line is: yes, if there are other ways of improving your workflow, that would be much preferable. :)

**Comment by PeterMarkuson** [comment]
Thank you very much for commenting back and explaining your point of view in detail. I am sorry for my noob point of view, but I was thinking you need to parse the source *.md files anyway to convert them to html. Also, I believed such parsing is required when building the TOC or dealing with the anchors. So my hope was you can re-use this feature and use it for enhancing the include tag easily. So, sorry to hear that this might not be done, but was worth a try :)

Regarding my use case: I am writing documentation for more generic topics first (e.g. networking) which makes sense to read/structure in a single page to have the whole context and to understand the topic easily by reading from top to bottom. This page on its own is useful. In another part of the documentation I write how-to guides which refer and use parts of the generic documentation and I use include for that.

Without the enhanced feature now I need to to the following:
1. Writing the atomic bits and pieces
2. Include all of them one after the other via include tag in the generic documentation
3. Include some of them in the right place of the guides.

The 1st step would not be required with the new functionality, but more than that there is another advantage: It's hard to edit the atomic bits and pieces, because it makes sense to read the generic documentation assembled in step 2 to understand and validate the content in the full context. When an update is required, I need to find and open the included atomic page and edit it there. Having it in one page would be easy to edit in place of the generic page, where the whole text is reviewed.

**Comment by benjaminwil** [comment]
Thanks for explaining further!

In my own career as a technical writer (which, it's worth mentioning, didn't include my writing very much generic or reusable documentation), much of the shareable pieces of writing ended up _not_ being shared, because it was too complicated to share it across a documentation project in a sophisticated way–even when using my team's preferred documentation writing software MadCap Flare (a piece of software I would recommend no one ever use!!!).

Based on what you've outlined, I think a custom macro for your project could be possible. But I don't know how we could formalize it into a proper feature yet. I would want to survey more people heavily using the `[[include]]` tag to find out what exactly this feature should look like.

That said, here is a [custom macro](https://github.com/gollum/gollum/wiki#custom-macros) that crudely implements what you're describing:

```
module Gollum
  class Macro
    class Snippet < Gollum::Macro
      def render(snippet_name)
        find_snippet(snippet_name)
      end

      private

      def find_snippet(snippet_name)
        snippet_body_regex = /\{{3}\(#{snippet_name}\)(.*)\}{3}/m

        if (match = snippets.match(snippet_body_regex))
          match.captures.first
        else
          nil
        end
      end

      def snippets
        @wiki.pages.detect { |page| page.name == "Snippets" }.text_data
      end
    end
  end
end
```

How does this work? If you have a page with the name "Snippets" (for me: `Snippets.md` in my root directory), you can define snippets in it like this:

```
{{{(snippet one)
Lorem ipsum dolor sit amet, consetetur sadipscing elitr,
sed diam nonumy eirmod tempor invidunt ut labore et dolore
magna aliquyam erat, sed diam voluptua. At vero eos et
accusam et justo duo dolores et ea rebum.
}}}

{{{(snippet two)
Another snippet!
}}}
```

And then, on other pages, call the snippet macro:

```
# Another Page.md

<<Snippet("snippet one")>>
```

On save, it loads the snippet.

**Comment by PeterMarkuson** [comment]
Thank you for your approach. I believe this would be helpful and I tried to implement it. When following your example (adding the two pages, copying the content into it and activating the macro): 

~I get the following error message on "Another page":~
~**Macro Error for Snippet: undefined method `text_data' for nil:NilClass**~

Update: After a second restart of Gollum the macro is not having an error anymore, but its pulling both snippets. The output on `another page` is now:

~~~
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. }}} {{{(snippet two) Another snippet!
~~~

So, the closing brackets of snippet one seem not to be recognized. 

Also markdown-formatting is not respected, so writing `*Lorem ipsum*` into the snippet to emphasize it, would result in `*Lorem ipsum*` on `another page` and not formatted bold as expected.

Also, on `Snippets.md` I see the `{{{(snippet one)` tag which should be hidden, right?

Also, if I get your example right, the name of the page to source the snippets from is hard-coded into the macro. So it's only looking on "Snippets", but not on any other pages? -> I have figured out how to make this with a variable :)

**Comment by benjaminwil** [comment]
Yes, this was absolutely a crude example I made in five minutes. And unfortunately my regular expression failed more than I thought it would. 😞 

But it sounds like you’re already tweaking this and making it more suitable for your use case? If you know what Markdown rendering gem your Gollum instance uses (the default is `kramdown`) I believe it’d be possible to build HTML output from your Markdown formatting, too. Check out `kramdown`’s documentation for information on how to do this.

**Comment by PeterMarkuson** [comment]
I started to customize it a bit, but my coding knowledge is very basic and I don't have any experience with Ruby, so it will be hard for me. I added an argument for the pagename to the macro and tried to adopt the regular expression to be more generic, but this is also a challenge for me. 

My current version is:
~~~
module Gollum
  class Macro
    class Snippet2 < Gollum::Macro
      def render(page_name, snippet_name, head_level)
        find_snippet(page_name, snippet_name, head_level)
      end

      private

      def find_snippet(page_name, snippet_name, head_level)
        snippet_body_regex = /^#{snippet_name}(.*)\#{#{head_level}}/m

        if (match = snippets(page_name).match(snippet_body_regex))
          match.captures.first
        else
          "[ERR] include not found!"
        end
      end

      def snippets(page_name)
        @wiki.pages.detect { |page| page.name == "#{page_name}" }.text_data
      end
    end
  end
end
~~~

Calling `<<Snippet2("Snippets","## First Heading", 2)>>` where "Snippets" is the pagename, "## First Heading" is the pattern to use as starting point and 2 is the heading-level to stop the include section. 

Example `Snippets` page
~~~
## First Heading
Here is some text

### Subheading
Additional Text

## Another Heading
Some more text
~~~

This delivers from the 2nd level heading `## First Heading` until (but not included) `## Another Heading`.  If the third parameter would be 3 instead of 2, the include will stop before `### Subheading`, because this is a 3rd level heading.

With this mechanism also no special patterns are required in the source text.

What is now missing for me is the conversion to html. I changed the markdown engine to github, because with kramdown I had issues with tables and unfortunately have no idea how to convert to html output.

So, thank you for the jumpstart with the Macro which clearly points me to the right direction and would be a suitable solution for me. Unfortunately my skills are not sufficient to develop a full solution on my own. I'll try to educate myself in Ruby, but if your time allows it and it's not too much effort for you, I would appreciate a little more help here.

**Comment by benjaminwil** [comment]
Conveniently, there is [one handy file in the `github-markup` gem's source code][1] shows us how we would take any input Markdown and turn it into HTML using any of the supported Markdown rendering gems.

It sounds like you're using the `github-markdown` gem. So it should just be a matter of wrapping your capture match in a `GitHub::Markdown.render` call (i.e. `GitHub::Markdown.render("My Markdown")`).

```ruby
module Gollum
  class Macro
    class Snippet < Gollum::Macro
      def render(page_name, snippet_name, head_level)
        find_snippet(page_name, snippet_name, head_level)
      end

      private

      def find_snippet(page_name, snippet_name, head_level)
        snippet_body_regex = /^#{snippet_name}(.*)\#{#{head_level}}/m

        if (match = snippets(page_name).match(snippet_body_regex))
          render_html(match.captures.first)
        else
          "[ERR] include not found!"
        end
      end

      def render_html(match)
        GitHub::Markdown.render(match)
      end

      def snippets(page_name)
        @wiki.pages.detect { |page| page.name == "#{page_name}" }.text_data
      end
    end
  end
end
```

Perhaps there is another set of arguments that are more ergonomic for you than `page_name`, `snippet_name`, and `head_level`. I'm sure someone who is more talented with regular expressions than I am could easily come up with one. That said: if this would already improve your workflow, I'm happy to hear it.

Because there is no special patterns required in the source text, it means if you ever come up with a different macro, or maybe would like a suite of macros that all do slightly different things, you're in luck.

[1]: https://github.com/github/markup/blob/v4.0.1/lib/github/markup/markdown.rb#L12-L30

**Comment by PeterMarkuson** [comment]
A big thank you from my side. I have managed to find a good solution with your helping hand. At the end I came up with the following Macro, which I will test now in my real world scenarios:

~~~ruby
# Purpose of this macro:
# Include a section from another page in place where the macro is called
# Limitations:
# Only working with GitHub::Markdown gem
#
# page_name             = Name of the Page without suffix (no .md)
# search_start          = A string to search for in the unrendered document
# search_end            = A string to stop searching or "" (empty) to continue until the end of the page
# include_start         = render the search_text (true) or exclude (false)
#
# The search strings need to be at the beginning of a line
#
# Example 1: <<IncludeSection("myPage", "# Introduction", "##", true)>>
# This will include from myPage the 1st level Heading "Introduction" and stops, when a 2nd level heading is found
#
# Example 2: <<IncludeSection("myPage", "## Epilog", "", false)>>
# This will include from myPage after the 2nd level Heading "Epilog" (but not rendering the heading) until the end of the page

module Gollum
  class Macro
    class IncludeSection < Gollum::Macro
      def render(page_name, search_start, search_end="", include_start=false)
        find(page_name, search_start, search_end, include_start)
      end

      private

      def find(page_name, search_start, search_end, include_start)
        if include_start
          regex = "(^#{search_start}"
        else
          regex = "^#{search_start}("
        end

        if search_end != ""
          regex += ".*?)(^#{search_end}\s)"
        else
          regex += ".*)"
        end

        rx = /#{regex}/mi
        if get_data(page_name) =~ rx
          render_html($1)
        else
          "[ERR] include not found with pattern<br>#{regex}"
        end
      end

      def render_html(match)
        GitHub::Markdown.render(match)
      end

      def get_data(page_name)
        @wiki.pages.detect { |page| page.name == "#{page_name}" }.text_data
      end
    end
  end
end
~~~

How does it work?

The user specifies a *name of the page* to look for as well as a *start string* and an *end string*. 
Condition: The strings shall be at the start of a line to avoid any false positives in running text. 

Everything between these two strings will be included. By default the start string is not included, but when it's a heading (like in many of my use case), it can be rendered by setting a parameter to true. With this I can select to include the original heading or define my own and just take the running text.

Also, if an empty *end string* is given, the page is rendered until the end of the page.

Function wise the magic is in the `def find(...)` where a proper regex is build depending on the arguments. If the regex is not successful, an error message with a hint what regex was used is given.

I am sure there is a more elegant solution. Also, I think it's not bulletproof. But for my needs it seem to work and if I find any improvements I will update it here.
